mm2src/coins/eth/eth_swap_v2/eth_maker_swap_v2.rs

```diff
@@ -0,0 +1,509 @@
+            tx_from_rpc,
+            maker_address,
+            maker_swap_v2_contract,
+            maker_status,
+            MakerPaymentStateV2::PaymentSent as u8,
+        )?;
+
+        let validation_args = {
+            let amount = wei_from_big_decimal(&args.amount, self.decimals)?;
+            MakerValidationArgs {
+                swap_id,
+                amount,
+                taker: self.my_addr().await,
+                taker_secret_hash: args.taker_secret_hash,
+                maker_secret_hash: args.maker_secret_hash,
+                payment_time_lock: args.time_lock,
+            }
+        };
+        match self.coin_type {
+            EthCoinType::Eth => {
+                let function = MAKER_SWAP_V2.function(ETH_MAKER_PAYMENT)?;
+                let decoded = decode_contract_call(function, &tx_from_rpc.input.0)?;
+                validate_eth_maker_payment_data(&decoded, &validation_args, function, tx_from_rpc.value)?;
+            },
+            EthCoinType::Erc20 { token_addr, .. } => {
+                let function = MAKER_SWAP_V2.function(ERC20_MAKER_PAYMENT)?;
+                let decoded = decode_contract_call(function, &tx_from_rpc.input.0)?;
+                validate_erc20_maker_payment_data(&decoded, &validation_args, function, token_addr)?;
+            },
+            EthCoinType::Nft { .. } => unreachable!(),
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn refund_maker_payment_v2_timelock_impl(
+        &self,
+        args: RefundMakerPaymentTimelockArgs<'_>,
+    ) -> Result<SignedEthTx, TransactionErr> {
+        let (token_address, gas_limit) = self
+            .gas_limit_v2
+            .gas_limit(
+                &self.coin_type,
+                EthPaymentType::MakerPayments,
+                PaymentMethod::RefundTimelock,
+            )
+            .map_err(|e| TransactionErr::Plain(ERRL!("{}", e)))?;
+
+        let maker_swap_v2_contract = self
+            .swap_v2_contracts
+            .as_ref()
+            .map(|contracts| contracts.maker_swap_v2_contract)
+            .ok_or_else(|| TransactionErr::Plain(ERRL!("Expected swap_v2_contracts to be Some, but found None")))?;
+        let payment_amount = try_tx_s!(wei_from_big_decimal(&args.amount, self.decimals));
+        let (maker_secret_hash, taker_secret_hash) = match args.tx_type_with_secret_hash {
+            SwapTxTypeWithSecretHash::MakerPaymentV2 {
+                maker_secret_hash,
+                taker_secret_hash,
+            } => (maker_secret_hash, taker_secret_hash),
+            _ => {
+                return Err(TransactionErr::Plain(ERRL!(
+                    "Unsupported swap tx type for timelock refund"
+                )))
+            },
+        };
+        let args = {
+            let taker_address = public_to_address(&Public::from_slice(args.taker_pub));
+            MakerRefundArgs {
+                payment_amount,
+                taker_address,
+                taker_secret: [0u8; 32],
```

Comment:
This can create confusion later, better to have different args for different refund types. What do you think?

Reply:
Agree, fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2211/commits/72ed3060ace61d9a14eca6fc5cedeb957ad0be54